movement/io/validators.py

```diff
@@ -1,39 +1,348 @@
-            logger.error(error_msg)
-            raise FileNotFoundError(error_msg)
-        return value
-
-    @field_validator("file_path")
-    def file_must_have_valid_suffix(cls, value):
-        if value.suffix not in (".h5", ".csv"):
-            error_msg = (
-                "Expected a file with pose estimation results from "
-                "DeepLabCut, in one of '.h5' or '.csv' formats. "
-                f"Received a file with suffix '{value.suffix}' instead."
-            )
-            logger.error(error_msg)
-            raise ValueError(error_msg)
-        return value
+    # Define class attributes
+    tracks_array: np.ndarray = field()
+    scores_array: Optional[np.ndarray] = field(default=None)
+    individual_names: Optional[List[str]] = field(
+        default=None,
+        converter=converters.optional(_list_of_str),
+    )
+    keypoint_names: Optional[List[str]] = field(
+        default=None,
+        converter=converters.optional(_list_of_str),
+    )
+    fps: Optional[float] = field(
+        default=None,
+        converter=converters.pipe(  # type: ignore
+            converters.optional(float), _set_fps_to_none_if_invalid
+        ),
+    )
+
+    # Add validators
+    @tracks_array.validator
+    def _validate_tracks_array(self, attribute, value):
+        _ensure_type_ndarray(value)
+        if value.ndim != 4:
+            log_and_raise_error(
+                ValueError,
+                f"Expected `{attribute}` to have 4 dimensions, "
+                f"but got {value.ndim}.",
+            )
+        if value.shape[-1] not in [2, 3]:
+            log_and_raise_error(
+                ValueError,
+                f"Expected `{attribute}` to have 2 or 3 spatial dimensions, "
+                f"but got {value.shape[-1]}.",
+            )
+
+    @scores_array.validator
+    def _validate_scores_array(self, attribute, value):
+        if value is not None:
+            _ensure_type_ndarray(value)
+            if value.shape != self.tracks_array.shape[:-1]:
+                log_and_raise_error(
+                    ValueError,
+                    f"Expected `{attribute}` to have shape "
+                    f"{self.tracks_array.shape[:-1]}, but got {value.shape}.",
+                )
+
+    @individual_names.validator
+    def _validate_individual_names(self, attribute, value):
+        if (value is not None) and (len(value) != self.tracks_array.shape[1]):
+            log_and_raise_error(
+                ValueError,
+                f"Expected {self.tracks_array.shape[1]} `{attribute}`, "
+                f"but got {len(value)}.",
+            )
+
+    @keypoint_names.validator
+    def _validate_keypoint_names(self, attribute, value):
+        if (value is not None) and (len(value) != self.tracks_array.shape[2]):
+            log_and_raise_error(
+                ValueError,
+                f"Expected {self.tracks_array.shape[2]} `{attribute}`, "
+                f"but got {len(value)}.",
+            )
```

Comment:
You are right, we have some duplication there. But your proposed solution won't work I think. The `_make_list_length_validator()` function needs to access `self` (because of `self.tracks_array.shape`), so it has to be a class method. Class methods are not available upon attribute initialisation, so we cannot do:
```python
individual_names: Optional[List[str]] = field(
        default=None,
        converter=converters.optional(_list_of_str),
        validator=self._make_list_length_validator(1),
    )
```
It seems to be an inherent limitation in how `attrs` internal work. I'll try to see whether I can find an alternative way of avoiding the duplication.

Reply:
An alternative approach would be to write a generic list length validator:
```python
def _validate_list_length(
    attribute: str, value: Optional[List], expected_length: int
):
    """Raise a ValueError if the list does not have the expected length."""
    if (value is not None) and (len(value) != expected_length):
        raise log_error(
            ValueError,
            f"Expected `{attribute}` to have length {expected_length}, "
            f"but got {len(value)}.",
        )
 ```
 Then the in-class validators become:
 ```python
 @individual_names.validator
    def _validate_individual_names(self, attribute, value):
        _validate_list_length(attribute, value, self.tracks_array.shape[1])

@keypoint_names.validator
    def _validate_keypoint_names(self, attribute, value):
        _validate_list_length(attribute, value, self.tracks_array.shape[2])
```

I kind of like this compromise, since it is compatible with `attrs` and it gives as a re-usable generic list length validator, which we will probably need in other places as well.

Reply:
The ``_make_list_length_validator()`` function does not need to be a class function nor does it need to access ``self`` as it returns the callable validator ``_validate_list_length`` in ``validator=_make_list_length_validator(1)``. This returned ``_validate_list_length`` will by default receive the same (``instance, attribute, value``) arguments as with the decorator approach. But I do agree that the alternative you suggested is easier to read.

Reply:
This is a neat trick! I will keep this in mind for future use, to get out of similarly tricky situations. 
For the case here, I will leave it as is now, in the interest of readability.